### ISSUE TITLE
Does not see `source.organizeImports` code action from typescript-language-server

### ISSUE THREAD

**Issue by maximbaz** [kak-lsp/kak-lsp#592]
I'm using https://github.com/typescript-language-server/typescript-language-server v0.9.6

Their README has [this part](https://github.com/typescript-language-server/typescript-language-server#code-actions-on-save):

> Server announces support for the following code action kinds:
> ...
> source.organizeImports.ts - organizes and removes unused imports
> ...
> Use like so:
```
"codeActionsOnSave": {
    "source.organizeImports.ts": true,
    // or just
    "source.organizeImports": true,
}
```

However `kak-lsp` doesn't seem to detect this code action, unless I'm doing something wrong.

Debug logs from `kak-lsp`:

```
Feb 17 18:03:14.132 DEBG From server: {"jsonrpc":"2.0","id":0,"result":{"capabilities":{"textDocumentSync":2,"completionProvider":{"triggerCharacters":[".","\"","'","/","@","<"],"resolveProvider":true},"codeActionProvider":{"codeActionKinds":["source.fixAll.ts","source.removeUnused.ts","source.addMissingImports.ts","source.organizeImports.ts"]},"definitionProvider":true,"documentFormattingProvider":true,"documentRangeFormattingProvider":true,"documentHighlightProvider":true,"documentSymbolProvider":true,"executeCommandProvider":{"commands":["_typescript.applyWorkspaceEdit","_typescript.applyCodeAction","_typescript.applyRefactoring","_typescript.organizeImports","_typescript.applyRenameFile"]},"hoverProvider":true,"renameProvider":true,"referencesProvider":true,"signatureHelpProvider":{"triggerCharacters":["(",",","<"]},"workspaceSymbolProvider":true,"implementationProvider":true,"typeDefinitionProvider":true,"foldingRangeProvider":true,"semanticTokensProvider":{"documentSelector":null,"legend":{"tokenTypes":["class","enum","interface","namespace","typeParameter","type","parameter","variable","enumMember","property","function","member"],"tokenModifiers":["declaration","static","async","readonly","defaultLibrary","local"]},"full":true,"range":true},"callsProvider":true}}}, module: kak_lsp::language_server_transport:151
```

Relevant part seems to only have `source.organizeImports.ts`, not the one without `.ts` suffix:

```json
      "codeActionProvider": {
        "codeActionKinds": [
          "source.fixAll.ts",
          "source.removeUnused.ts",
          "source.addMissingImports.ts",
          "source.organizeImports.ts"
        ]
      },
```

At first I thought this is the root cause of the bug, and a quick grep in `kak-lsp` found only one occurrence:

https://github.com/kak-lsp/kak-lsp/blob/88f52f0c6271df372942b8344ea2bbca43c58b25/src/general.rs#L206

I tried to append `"source.organizeImports.ts"` to the list, but it didn't seem to help, so I'm wondering if the bug is elsewhere.

I also saw that maybe we can make a workaround command to send this command manually?

https://www.reddit.com/r/neovim/comments/lwz8l7/how_to_use_tsservers_organize_imports_with_nvim/gpkueno/?context=3

What do you think is truly going on? 🙂

**Comment by krobelus** [comment]
Right, typescript-language-server (actually this behavior is inherited from tsserver) uses a special LSP command `_typescript.organizeImports`. Kind of defeats the point of LSP but OK.

> I also saw that maybe we can make a workaround command to send this command manually?
> https://www.reddit.com/r/neovim/comments/lwz8l7/how_to_use_tsservers_organize_imports_with_nvim/gpkueno/?context=3

That was very helpful; this command should be equivalent

```kak
define-command tsserver-organize-imports -docstring "Ask the typescript language server to organize imports in the buffer" %{
    lsp-execute-command _typescript.organizeImports """[\""%val{buffile}\""]"""
}
```

We should probably add this to rc/lsp.kak.

Due to a bug in kak-lsp, a version with lsp-execute-command-sync won't work yet.

---

I don't understand why they don't just use the generic `textDocument/codeAction`.
gopls behaves better here, they just use the generic `textDocument/codeAction`.
Technically, code actions are dependent on the cursor position, but gopls simply sends "Organize imports" for any cursor position.

Interestingly, typescript-language-server does send generic code actions for removing unused imports (and probably others too), but only if the cursor is on the import statement.
I tried with

```
echo "import { x } from './lib';
" > test.ts
echo "export declare const x = 1;" > lib.ts
```

with the cursor on the first line, there is a code action for removing the import, but not with the cursor on the second line.

**Comment by krobelus** [comment]
So this feature is not exposed as code action but only as "workspace command".
I didn't check how the codeActionsOnSave is interpreted by VSCode's typescript support, but probably they special case the organizeImports code action and send the workspace command instead.

**Comment by maximbaz** [comment]
Thanks for looking into this! What you described makes sense.

I confirm the command works 👍 

I don't know much about typescript but I think I read in some vscode issue that typescript-language-server was invented before LSP was a thing, so they had their own protocol and everything, and are still in progress of migrating to be a proper LSP now that it's a standard. I suspect they would be happy for issues describing exactly what they are missing or doing wrong or bad compared to other LSP implementations..
